roofit/roofitcore/inc/LinkDef.h

```diff
@@ -29,12 +29,21 @@
 #pragma link C++ class RooAbsBinning- ;
 #pragma link C++ class RooAbsCategory+ ;
 #pragma read sourceClass="RooAbsCategory" targetClass="RooAbsCategory" version="[1]" \
-  source="TObjArray _types" target="_types" \
-  code="{TObject* obj; TIterator* it = onfile._types.MakeIterator();\
-         while ((obj=it->Next())) {\
-           auto cat = dynamic_cast<const RooCatType*>(obj); assert(cat);\
-           _types.push_back(new RooCatType(*cat)); }\
-         delete it; }";
+  include="RooFitLegacy/RooCatTypeLegacy.h" \
+  source="TObjArray _types" target="_stateNames,_insertionOrder" \
+  code="{for (const auto* obj : onfile._types) { \
+           auto catType = dynamic_cast<const RooCatType*>(obj); assert(catType); \
+           _stateNames[catType->GetName()] = catType->getVal(); \
+           _insertionOrder.push_back(catType->getVal()); \
+         }}";
+#pragma read sourceClass="RooAbsCategory" targetClass="RooAbsCategory" version="[2]" \
+  include="RooFitLegacy/RooCatTypeLegacy.h" \
+  source="std::vector<RooCatType*> _types" target="_stateNames,_insertionOrder" \
+  code="{for (const auto catType : onfile._types) { _stateNames[catType->GetName()] = catType->getVal();\
+                                                    _insertionOrder.push_back(catType->getVal());\
+                                                  } }";
+#pragma read sourceClass="RooAbsCategory" targetClass="RooAbsCategory" version="[1-2]" include="RooFitLegacy/RooCatTypeLegacy.h" \
+  source="RooCatType _value" target="_currentIndex" code="{ _currentIndex = onfile._value.getVal(); }"
```

Comment:
just curious..what on earth does this do?

Reply:
It's schema evolution. These are read rules telling ROOT how to read an old class version on file that stores category states in the 284-bytes-too-large `RooCatType`. You basically say:
Get me the RooCatType from file, and I do with it what's necessary.